// @antv/g2plot/lib/plots/tiny-line/adaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptor = void 0;
var util_1 = require("@antv/util");
var utils_1 = require("../../utils");
var common_1 = require("../../adaptor/common");
var geometries_1 = require("../../adaptor/geometries");
/**
 * 字段
 * @param params
 */
function geometry(params) {
    var chart = params.chart, options = params.options;
    var data = options.data, color = options.color, lineStyle = options.lineStyle, pointMapping = options.point;
    var seriesData = data.map(function (y, x) {
        return { x: x, y: y };
    });
    chart.data(seriesData);
    // line geometry 处理
    var p = util_1.deepMix({}, params, {
        options: {
            xField: 'x',
            yField: 'y',
            line: {
                color: color,
                style: lineStyle,
            },
            point: pointMapping,
        },
    });
    geometries_1.line(p);
    geometries_1.point(p);
    chart.axis(false);
    chart.legend(false);
    return params;
}
/**
 * 迷你折线图适配器
 * @param chart
 * @param options
 */
function adaptor(params) {
    return utils_1.flow(geometry, common_1.scale({}), common_1.theme, common_1.tooltip, common_1.animation, common_1.annotation())(params);
}
exports.adaptor = adaptor;
