// @antv/g2plot/lib/plots/gauge/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Gauge = void 0;
var tslib_1 = require("tslib");
var plot_1 = require("../../core/plot");
var adaptor_1 = require("./adaptor");
var constant_1 = require("./constant");
// 注册 shape
require("./shapes/gauge");
/**
 * 仪表盘盘
 */
var Gauge = /** @class */ (function (_super) {
    tslib_1.__extends(Gauge, _super);
    function Gauge() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'gauge';
        return _this;
    }
    Gauge.prototype.getDefaultOptions = function (options) {
        var _a;
        var percent = options.percent;
        return {
            percent: 0,
            range: {
                ticks: [0, percent, 1],
            },
            innerRadius: 0.9,
            radius: 0.95,
            startAngle: (-7 / 6) * Math.PI,
            endAngle: (1 / 6) * Math.PI,
            syncViewPadding: true,
            axis: {
                line: null,
                label: {
                    offset: -24,
                    style: {
                        textAlign: 'center',
                        textBaseline: 'middle',
                    },
                },
                subTickLine: {
                    length: -8,
                },
                tickLine: {
                    length: -12,
                },
                grid: null,
            },
            indicator: {
                pointer: {
                    style: {
                        stroke: '#1890FF',
                        lineWidth: 5,
                        lineCap: 'round',
                    },
                },
                pin: {
                    style: {
                        stroke: '#1890FF',
                        r: 9.75,
                        lineWidth: 4.5,
                        fill: '#fff',
                    },
                },
            },
            statistic: {
                title: false,
                content: {
                    formatter: function (_a) {
                        var percent = _a.percent;
                        return (percent * 100).toFixed(2) + "%";
                    },
                    style: {
                        opacity: 0.75,
                        fontSize: 30,
                        textAlign: 'center',
                        textBaseline: 'bottom',
                    },
                    offsetY: -16,
                },
            },
            meta: (_a = {},
                // 两个 view 的 scale 同步到 v 上
                _a[constant_1.RANGE_VALUE] = {
                    sync: 'v',
                },
                _a[constant_1.PERCENT] = {
                    sync: 'v',
                },
                _a),
        };
    };
    /**
     * 获取适配器
     */
    Gauge.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return Gauge;
}(plot_1.Plot));
exports.Gauge = Gauge;
