// @antv/g2plot/esm/plots/bullet/adaptor.js
import { deepMix, isNumber } from '@antv/util';
import { interaction, animation, theme, tooltip } from '../../adaptor/common';
import { flow, pick } from '../../utils';
import { AXIS_META_CONFIG_KEYS } from '../../constant';
import { interval, point } from '../../adaptor/geometries';
import { transformData } from './utils';
/**
 * geometry 处理
 * @param params
 */
function geometry(params) {
    var _a;
    var chart = params.chart, options = params.options;
    var style = options.style, targetField = options.targetField, rangeField = options.rangeField, measureField = options.measureField, xField = options.xField, color = options.color, layout = options.layout, size = options.size;
    // 处理数据
    var _b = transformData(options), min = _b.min, max = _b.max, ds = _b.ds;
    // 需要统一比列尺
    chart.scale((_a = {},
        _a[measureField] = {
            min: min,
            max: max,
        },
        _a[rangeField] = {
            sync: "" + measureField,
        },
        _a[targetField] = {
            sync: "" + measureField,
        },
        _a));
    chart.data(ds);
    chart.axis("" + rangeField, false);
    chart.axis("" + targetField, false);
    // rangeGeometry
    var r = deepMix({}, params, {
        options: {
            xField: xField,
            yField: rangeField,
            seriesField: 'rKey',
            isStack: true,
            interval: {
                color: color === null || color === void 0 ? void 0 : color.range,
                style: style === null || style === void 0 ? void 0 : style.range,
                size: size === null || size === void 0 ? void 0 : size.range,
            },
        },
    });
    interval(r);
    // 范围值的 tooltip 隐藏掉
    chart.geometries[0].tooltip(false);
    // measureGeometry
    var m = deepMix({}, params, {
        options: {
            xField: xField,
            yField: measureField,
            seriesField: 'mKey',
            isStack: true,
            interval: {
                color: color === null || color === void 0 ? void 0 : color.measure,
                style: style === null || style === void 0 ? void 0 : style.measure,
                size: size === null || size === void 0 ? void 0 : size.measure,
            },
        },
    });
    interval(m);
    // targetGeometry
    var t = deepMix({}, params, {
        options: {
            xField: xField,
            yField: targetField,
            seriesField: 'tKey',
            point: {
                color: color === null || color === void 0 ? void 0 : color.target,
                style: style === null || style === void 0 ? void 0 : style.target,
                size: isNumber(size === null || size === void 0 ? void 0 : size.target) ? Number(size.target) / 2 : size.target,
                shape: layout === 'horizontal' ? 'line' : 'hyphen',
            },
        },
    });
    point(t);
    // 水平的时候，要转换坐标轴
    if (layout === 'horizontal') {
        chart.coordinate().transpose();
    }
    return params;
}
/**
 * meta 配置
 * @param params
 */
function meta(params) {
    var _a;
    var chart = params.chart, options = params.options;
    var xAxis = options.xAxis, yAxis = options.yAxis, meta = options.meta, targetField = options.targetField, rangeField = options.rangeField, measureField = options.measureField, xField = options.xField;
    if (meta) {
        var scales = deepMix({}, meta, (_a = {},
            _a[xField] = pick(xAxis, AXIS_META_CONFIG_KEYS),
            _a[measureField] = pick(yAxis, AXIS_META_CONFIG_KEYS),
            _a[targetField] = {
                sync: "" + measureField,
            },
            _a[rangeField] = {
                sync: "" + measureField,
            },
            _a));
        chart.scale(scales);
    }
    return params;
}
/**
 * axis 配置
 * @param params
 */
function axis(params) {
    var chart = params.chart, options = params.options;
    var xAxis = options.xAxis, yAxis = options.yAxis, xField = options.xField, measureField = options.measureField;
    // 为 false 则是不显示轴
    if (xAxis === false) {
        chart.axis("" + xField, false);
    }
    else {
        chart.axis("" + xField, xAxis);
    }
    if (yAxis === false) {
        chart.axis("" + measureField, false);
    }
    else {
        chart.axis("" + measureField, yAxis);
    }
    return params;
}
/**
 * legend 配置
 * @param params
 */
function legend(params) {
    var chart = params.chart, options = params.options;
    var legend = options.legend;
    chart.removeInteraction('legend-filter');
    // @TODO 后续看是否内部自定义一个 legend
    chart.legend(legend);
    // 默认关闭掉所在 color 字段的 legend, 从而不影响自定义的legend
    chart.legend('rKey', false);
    chart.legend('mKey', false);
    chart.legend('tKey', false);
    return params;
}
/**
 * label 配置
 * @param params
 */
function label(params) {
    var chart = params.chart, options = params.options;
    var label = options.label, measureField = options.measureField, targetField = options.targetField, rangeField = options.rangeField;
    var _a = chart.geometries, rangeGeometry = _a[0], measureGeometry = _a[1], targetGeometry = _a[2];
    if (label === null || label === void 0 ? void 0 : label.range) {
        rangeGeometry.label("" + rangeField, label.range);
    }
    if (label === null || label === void 0 ? void 0 : label.measure) {
        measureGeometry.label("" + measureField, label.measure);
    }
    if (label === null || label === void 0 ? void 0 : label.target) {
        targetGeometry.label("" + targetField, label.target);
    }
    return params;
}
/**
 * 子弹图适配器
 * @param chart
 * @param options
 */
export function adaptor(params) {
    // flow 的方式处理所有的配置到 G2 API
    flow(geometry, meta, axis, legend, theme, label, tooltip, interaction, animation)(params);
}
