// @antv/g2plot/esm/plots/pie/index.js
import { __extends } from "tslib";
import { deepMix } from '@antv/util';
import { Plot } from '../../core/plot';
import { adaptor } from './adaptor';
import './interaction';
import './label';
var Pie = /** @class */ (function (_super) {
    __extends(Pie, _super);
    function Pie() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'pie';
        return _this;
    }
    /**
     * 获取 饼图 默认配置项
     */
    Pie.prototype.getDefaultOptions = function () {
        return deepMix({}, _super.prototype.getDefaultOptions.call(this), {
            legend: {
                position: 'right',
            },
            tooltip: {
                shared: false,
                showTitle: false,
            },
            /** 饼图样式, 不影响暗黑主题 */
            pieStyle: {
                stroke: 'white',
                lineWidth: 1,
            },
            /** 饼图中心文本默认样式 */
            statistic: {
                title: {
                    style: { fontSize: 14, fontWeight: 300, fill: '#4D4D4D', textAlign: 'center' },
                },
                content: {
                    style: { fontSize: 21, fontWeight: 'bold', fill: '#4D4D4D', textAlign: 'center' },
                },
            },
        });
    };
    /**
     * 获取 饼图 的适配器
     */
    Pie.prototype.getSchemaAdaptor = function () {
        return adaptor;
    };
    return Pie;
}(Plot));
export { Pie };
