// @antv/g2plot/lib/plots/line/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Line = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var plot_1 = require("../../core/plot");
var adaptor_1 = require("./adaptor");
var Line = /** @class */ (function (_super) {
    tslib_1.__extends(Line, _super);
    function Line() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'line';
        return _this;
    }
    /**
     * 获取 折线图 默认配置
     */
    Line.prototype.getDefaultOptions = function (options) {
        var _a;
        var xField = options.xField;
        return util_1.deepMix({}, _super.prototype.getDefaultOptions.call(this), {
            tooltip: {
                showMarkers: true,
                showCrosshairs: true,
                crosshairs: {
                    type: 'x',
                },
            },
            meta: (_a = {},
                _a[xField] = {
                    range: [0, 1],
                },
                _a),
        });
    };
    /**
     * 获取 折线图 的适配器
     */
    Line.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return Line;
}(plot_1.Plot));
exports.Line = Line;
