// @antv/g2plot/lib/core/plot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Plot = void 0;
var tslib_1 = require("tslib");
var g2_1 = require("@antv/g2");
var util_1 = require("@antv/util");
var event_emitter_1 = tslib_1.__importDefault(require("@antv/event-emitter"));
var size_sensor_1 = require("size-sensor");
var utils_1 = require("../utils");
/**
 * 所有 plot 的基类
 */
var Plot = /** @class */ (function (_super) {
    tslib_1.__extends(Plot, _super);
    function Plot(container, options) {
        var _this = _super.call(this) || this;
        /** plot 类型名称 */
        _this.type = 'base';
        _this.container = typeof container === 'string' ? document.getElementById(container) : container;
        _this.options = util_1.deepMix({}, _this.getDefaultOptions(options), options);
        _this.createG2();
        _this.bindEvents();
        return _this;
    }
    /**
     * 创建 G2 实例
     */
    Plot.prototype.createG2 = function () {
        var _a = this.options, width = _a.width, height = _a.height, padding = _a.padding, appendPadding = _a.appendPadding, renderer = _a.renderer, pixelRatio = _a.pixelRatio, syncViewPadding = _a.syncViewPadding;
        this.chart = new g2_1.Chart(tslib_1.__assign(tslib_1.__assign({ container: this.container, autoFit: false }, this.getChartSize(width, height)), { padding: padding,
            appendPadding: appendPadding,
            renderer: renderer,
            pixelRatio: pixelRatio, localRefresh: false, // 默认关闭，目前 G 还有一些位置问题，难以排查！
            syncViewPadding: syncViewPadding }));
    };
    /**
     * 计算默认的 chart 大小。逻辑简化：如果存在 width 或 height，则直接使用，否则使用容器大小
     * @param width
     * @param height
     */
    Plot.prototype.getChartSize = function (width, height) {
        var chartSize = utils_1.getContainerSize(this.container);
        return { width: width || chartSize.width, height: height || chartSize.height };
    };
    /**
     * 绑定代理所有 G2 的事件
     */
    Plot.prototype.bindEvents = function () {
        var _this = this;
        if (this.chart) {
            this.chart.on('*', function (e) {
                if (e === null || e === void 0 ? void 0 : e.type) {
                    _this.emit(e.type, e);
                }
            });
        }
    };
    /**
     * 获取默认的 options 配置项
     * 每个组件都可以复写
     */
    Plot.prototype.getDefaultOptions = function (options) {
        return {
            renderer: 'canvas',
            tooltip: {
                shared: true,
                showMarkers: false,
                offset: 20,
            },
            xAxis: {
                nice: true,
                label: {
                    autoRotate: true,
                    autoHide: true,
                },
            },
            yAxis: {
                nice: true,
                label: {
                    autoHide: true,
                    autoRotate: false,
                },
            },
            animation: true,
        };
    };
    /**
     * 绘制
     */
    Plot.prototype.render = function () {
        // 暴力处理，先清空再渲染，需要 G2 层自行做好更新渲染
        this.chart.clear();
        // 执行 adaptor
        this.execAdaptor();
        // 渲染
        this.chart.render();
        // 绑定
        this.bindSizeSensor();
    };
    /**
     * 更新配置
     * @param options
     */
    Plot.prototype.update = function (options) {
        this.options = options;
        this.render();
    };
    /**
     * 设置状态
     * @param type 状态类型，支持 'active' | 'inactive' | 'selected' 三种
     * @param conditions 条件，支持数组
     * @param status 是否激活，默认 true
     */
    Plot.prototype.setState = function (type, condition, status) {
        if (status === void 0) { status = true; }
        var elements = utils_1.getAllElements(this.chart);
        util_1.each(elements, function (ele) {
            if (condition(ele.getData())) {
                ele.setState(type, status);
            }
        });
    };
    /**
     * 获取状态
     */
    Plot.prototype.getStates = function () {
        var elements = utils_1.getAllElements(this.chart);
        var stateObjects = [];
        util_1.each(elements, function (element) {
            var data = element.getData();
            var states = element.getStates();
            util_1.each(states, function (state) {
                stateObjects.push({ data: data, state: state, geometry: element.geometry, element: element });
            });
        });
        return stateObjects;
    };
    /**
     * 更新数据
     * @param options
     */
    Plot.prototype.changeData = function (data) {
        // 临时方案，会在 G2 做处理
        this.update(tslib_1.__assign(tslib_1.__assign({}, this.options), { data: data }));
        // this.chart.changeData(data);
    };
    /**
     * 修改画布大小
     * @param width
     * @param height
     */
    Plot.prototype.changeSize = function (width, height) {
        this.chart.changeSize(width, height);
    };
    /**
     * 销毁
     */
    Plot.prototype.destroy = function () {
        // 取消 size-sensor 的绑定
        this.unbindSizeSensor();
        // G2 的销毁
        this.chart.destroy();
        // 清空已经绑定的事件
        this.off();
    };
    /**
     * 执行 adaptor 操作
     */
    Plot.prototype.execAdaptor = function () {
        var adaptor = this.getSchemaAdaptor();
        var padding = this.options.padding;
        // 更新 padding
        this.chart.padding = padding;
        // 转化成 G2 API
        adaptor({
            chart: this.chart,
            options: this.options,
        });
    };
    /**
     * 当图表容器大小变化的时候，执行的函数
     */
    Plot.prototype.triggerResize = function () {
        this.chart.forceFit();
    };
    /**
     * 绑定 dom 容器大小变化的事件
     */
    Plot.prototype.bindSizeSensor = function () {
        var _this = this;
        if (this.unbind) {
            return;
        }
        var _a = this.options.autoFit, autoFit = _a === void 0 ? true : _a;
        if (autoFit) {
            this.unbind = size_sensor_1.bind(this.container, function () {
                _this.triggerResize();
            });
        }
    };
    /**
     * 取消绑定
     */
    Plot.prototype.unbindSizeSensor = function () {
        if (this.unbind) {
            this.unbind();
            this.unbind = undefined;
        }
    };
    return Plot;
}(event_emitter_1.default));
exports.Plot = Plot;
