// @antv/g2plot/lib/plots/funnel/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Funnel = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var plot_1 = require("../../core/plot");
var adaptor_1 = require("./adaptor");
var Funnel = /** @class */ (function (_super) {
    tslib_1.__extends(Funnel, _super);
    function Funnel() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'funnel';
        return _this;
    }
    /**
     * 获取 漏斗图 默认配置项
     */
    Funnel.prototype.getDefaultOptions = function (options) {
        var compareField = options.compareField;
        return util_1.deepMix({}, _super.prototype.getDefaultOptions.call(this), {
            annotation: compareField ? function (xField, yField) { return "" + yField; } : function (xField, yField) { return xField + " " + yField; },
            label: {
                callback: function (xField, yField, percent) {
                    return {
                        content: "\u8F6C\u5316\u7387" + +percent * 100 + "%",
                    };
                },
            },
            tooltip: {
                showCrosshairs: false,
            },
        });
    };
    /**
     * 获取 漏斗图 的适配器
     */
    Funnel.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return Funnel;
}(plot_1.Plot));
exports.Funnel = Funnel;
