// @antv/g2plot/lib/plots/bar/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bar = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var percent_tooltip_1 = require("../../utils/percent-tooltip");
var plot_1 = require("../../core/plot");
var adaptor_1 = require("./adaptor");
/**
 * 条形图
 */
var Bar = /** @class */ (function (_super) {
    tslib_1.__extends(Bar, _super);
    function Bar() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'bar';
        return _this;
    }
    /**
     * 获取 条形图 默认配置
     */
    Bar.prototype.getDefaultOptions = function (options) {
        var isPercent = options.isPercent, isRange = options.isRange, label = options.label, xField = options.xField;
        return util_1.deepMix({}, _super.prototype.getDefaultOptions.call(this), {
            label: label && isRange
                ? tslib_1.__assign({ content: function (item) {
                        var _a;
                        return (_a = item[xField]) === null || _a === void 0 ? void 0 : _a.join('-');
                    } }, label) : label,
            tooltip: {
                shared: true,
                showMarkers: false,
                offset: 20,
                customContent: isPercent
                    ? function (value, items) {
                        return percent_tooltip_1.getPercentTooltipTemplate(value, items);
                    }
                    : '',
            },
            interactions: [{ type: 'active-region' }],
        });
    };
    /**
     * 获取 条形图 的适配器
     */
    Bar.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return Bar;
}(plot_1.Plot));
exports.Bar = Bar;
