// @antv/g2plot/esm/adaptor/geometries/interval.js
import { __assign } from "tslib";
import { deepMix, isNil } from '@antv/util';
import { geometry } from './base';
/**
 * 柱形图其他的 adaptor
 * @param params
 */
function otherAdaptor(params) {
    var chart = params.chart, options = params.options, ext = params.ext;
    var seriesField = options.seriesField, isGroup = options.isGroup, isStack = options.isStack, marginRatio = options.marginRatio, widthRatio = options.widthRatio;
    var g = ext.geometry;
    /**
     * adjust
     */
    if (seriesField) {
        // group
        if (isGroup) {
            g.adjust({
                type: 'dodge',
                marginRatio: marginRatio,
            });
        }
        else if (isStack) {
            // stack
            g.adjust({
                type: 'stack',
                marginRatio: marginRatio,
            });
        }
    }
    // widthRatio
    if (!isNil(widthRatio)) {
        chart.theme({
            columnWidthRatio: widthRatio,
        });
    }
    return params;
}
export function interval(params) {
    var options = params.options;
    var interval = options.interval, seriesField = options.seriesField;
    // 保障一定要存在 interval 映射
    var ext = (interval
        ? geometry(deepMix({}, params, {
            options: {
                type: 'interval',
                colorField: seriesField,
                mapping: interval,
            },
        }))
        : params).ext;
    return otherAdaptor(__assign(__assign({}, params), { ext: ext }));
}
