// @antv/g2plot/lib/plots/word-cloud/adaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptor = void 0;
var common_1 = require("../../adaptor/common");
var utils_1 = require("../../utils");
var utils_2 = require("./utils");
/**
 * geometry 配置处理
 * @param params
 */
function geometry(params) {
    var chart = params.chart;
    var data = utils_2.transform(params);
    chart.data(data);
    chart.point().position('x*y').shape('word-cloud');
    return params;
}
/**
 * color 配置处理
 * @param params
 */
function color(params) {
    var chart = params.chart, options = params.options;
    var wordField = options.wordField, color = options.color;
    var geometry = utils_1.findGeometry(chart, 'point');
    geometry.color(wordField, color);
    return params;
}
/**
 * meta 配置
 * @param params
 */
function meta(params) {
    return utils_1.flow(common_1.scale({
        x: { nice: false },
        y: { nice: false },
    }))(params);
}
/**
 * coord 配置
 * @param params
 */
function coord(params) {
    var chart = params.chart;
    chart.coordinate().reflect('y');
    return params;
}
/**
 * axis 配置
 * 词云图不显示轴信息
 * @param params
 */
function axis(params) {
    var chart = params.chart;
    chart.axis(false);
    return params;
}
/**
 * label 配置
 * 词云图不显示 label 信息
 * @param params
 */
function label(params) {
    var chart = params.chart;
    var geometry = utils_1.findGeometry(chart, 'point');
    geometry.label(false);
    return params;
}
/**
 * legend 配置
 * 词云图不显示 legend 信息
 * @param params
 */
function legend(params) {
    var chart = params.chart;
    chart.legend(false);
    return params;
}
/**
 * 词云图适配器
 * @param chart
 * @param options
 */
function adaptor(params) {
    // flow 的方式处理所有的配置到 G2 API
    utils_1.flow(geometry, meta, coord, axis, label, color, legend, common_1.tooltip, common_1.interaction, common_1.animation, common_1.theme, common_1.state)(params);
}
exports.adaptor = adaptor;
