// @antv/g2plot/esm/plots/tiny-area/adaptor.js
import { deepMix } from '@antv/util';
import { theme, scale, animation, annotation, tooltip } from '../../adaptor/common';
import { flow } from '../../utils';
import { area, line, point } from '../../adaptor/geometries';
/**
 * 字段
 * @param params
 */
function geometry(params) {
    var chart = params.chart, options = params.options;
    var data = options.data, color = options.color, areaStyle = options.areaStyle, pointOptions = options.point, lineOptions = options.line;
    var seriesData = data.map(function (y, x) {
        return { x: x, y: y };
    });
    chart.data(seriesData);
    var p = deepMix({}, params, {
        options: {
            xField: 'x',
            yField: 'y',
            area: { color: color, style: areaStyle },
            line: lineOptions,
            point: pointOptions,
        },
    });
    // area geometry 处理
    area(p);
    line(p);
    point(p);
    chart.axis(false);
    chart.legend(false);
    return params;
}
/**
 * 迷你面积图适配器
 * @param chart
 * @param options
 */
export function adaptor(params) {
    return flow(geometry, scale({}), tooltip, theme, animation, annotation())(params);
}
