// @antv/g2plot/lib/plots/scatter/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scatter = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var plot_1 = require("../../core/plot");
var adaptor_1 = require("./adaptor");
require("./interaction");
var Scatter = /** @class */ (function (_super) {
    tslib_1.__extends(Scatter, _super);
    function Scatter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 图表类型 */
        _this.type = 'point';
        return _this;
    }
    /**
     * 获取散点图的适配器
     */
    Scatter.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    Scatter.prototype.getDefaultOptions = function (options) {
        var shapeField = options.shapeField, colorField = options.colorField, legend = options.legend;
        return util_1.deepMix({}, _super.prototype.getDefaultOptions.call(this), {
            size: 4,
            /** pointStyle 跟随主题默认样式 */
            tooltip: {
                shared: null,
                showTitle: false,
            },
            /**
             * legend 没有指定时根据 shapeField 和 colorField 来设置默认值
             */
            legend: util_1.isBoolean(legend) ? legend : legend || !!(shapeField || colorField),
        });
    };
    return Scatter;
}(plot_1.Plot));
exports.Scatter = Scatter;
