// @antv/g2plot/esm/plots/tiny-column/adaptor.js
import { deepMix } from '@antv/util';
import { theme, scale, animation, annotation, tooltip } from '../../adaptor/common';
import { flow } from '../../utils';
import { interval } from '../../adaptor/geometries';
/**
 * 字段
 * @param params
 */
function geometry(params) {
    var chart = params.chart, options = params.options;
    var data = options.data, color = options.color, columnStyle = options.columnStyle, columnWidthRatio = options.columnWidthRatio;
    var seriesData = data.map(function (y, x) {
        return { x: x, y: y };
    });
    chart.data(seriesData);
    var p = deepMix({}, params, {
        options: {
            xField: 'x',
            yField: 'y',
            widthRatio: columnWidthRatio,
            interval: {
                style: columnStyle,
                color: color,
            },
        },
    });
    interval(p);
    chart.axis(false);
    chart.legend(false);
    return params;
}
/**
 * 迷你柱形图适配器
 * @param chart
 * @param options
 */
export function adaptor(params) {
    return flow(geometry, scale({}), tooltip, theme, animation, annotation())(params);
}
