// @antv/g2plot/lib/plots/rose/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rose = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var plot_1 = require("../../core/plot");
var adaptor_1 = require("./adaptor");
var Rose = /** @class */ (function (_super) {
    tslib_1.__extends(Rose, _super);
    function Rose() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /** 玫瑰图 */
        _this.type = 'rose';
        return _this;
    }
    /**
     * 获取默认的 options 配置项
     */
    Rose.prototype.getDefaultOptions = function () {
        return util_1.deepMix({}, _super.prototype.getDefaultOptions.call(this), {
            xAxis: false,
            yAxis: false,
            legend: {
                position: 'right',
                offsetX: -10,
            },
            sectorStyle: {
                stroke: '#fff',
                lineWidth: 1,
            },
        });
    };
    /**
     * 获取 玫瑰图 的适配器
     */
    Rose.prototype.getSchemaAdaptor = function () {
        return adaptor_1.adaptor;
    };
    return Rose;
}(plot_1.Plot));
exports.Rose = Rose;
