// @antv/g2plot/lib/adaptor/geometries/interval.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.interval = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var base_1 = require("./base");
/**
 * 柱形图其他的 adaptor
 * @param params
 */
function otherAdaptor(params) {
    var chart = params.chart, options = params.options, ext = params.ext;
    var seriesField = options.seriesField, isGroup = options.isGroup, isStack = options.isStack, marginRatio = options.marginRatio, widthRatio = options.widthRatio;
    var g = ext.geometry;
    /**
     * adjust
     */
    if (seriesField) {
        // group
        if (isGroup) {
            g.adjust({
                type: 'dodge',
                marginRatio: marginRatio,
            });
        }
        else if (isStack) {
            // stack
            g.adjust({
                type: 'stack',
                marginRatio: marginRatio,
            });
        }
    }
    // widthRatio
    if (!util_1.isNil(widthRatio)) {
        chart.theme({
            columnWidthRatio: widthRatio,
        });
    }
    return params;
}
function interval(params) {
    var options = params.options;
    var interval = options.interval, seriesField = options.seriesField;
    // 保障一定要存在 interval 映射
    var ext = (interval
        ? base_1.geometry(util_1.deepMix({}, params, {
            options: {
                type: 'interval',
                colorField: seriesField,
                mapping: interval,
            },
        }))
        : params).ext;
    return otherAdaptor(tslib_1.__assign(tslib_1.__assign({}, params), { ext: ext }));
}
exports.interval = interval;
