// @antv/g2plot/esm/plots/tiny-line/adaptor.js
import { deepMix } from '@antv/util';
import { flow } from '../../utils';
import { scale, theme, animation, annotation, tooltip } from '../../adaptor/common';
import { line, point } from '../../adaptor/geometries';
/**
 * 字段
 * @param params
 */
function geometry(params) {
    var chart = params.chart, options = params.options;
    var data = options.data, color = options.color, lineStyle = options.lineStyle, pointMapping = options.point;
    var seriesData = data.map(function (y, x) {
        return { x: x, y: y };
    });
    chart.data(seriesData);
    // line geometry 处理
    var p = deepMix({}, params, {
        options: {
            xField: 'x',
            yField: 'y',
            line: {
                color: color,
                style: lineStyle,
            },
            point: pointMapping,
        },
    });
    line(p);
    point(p);
    chart.axis(false);
    chart.legend(false);
    return params;
}
/**
 * 迷你折线图适配器
 * @param chart
 * @param options
 */
export function adaptor(params) {
    return flow(geometry, scale({}), theme, tooltip, animation, annotation())(params);
}
