// @antv/g2plot/lib/plots/ring-progress/adaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptor = void 0;
var utils_1 = require("../../utils");
var common_1 = require("../../adaptor/common");
var adaptor_1 = require("../progress/adaptor");
/**
 * coordinate 配置
 * @param params
 */
function coordinate(params) {
    var chart = params.chart, options = params.options;
    var innerRadius = options.innerRadius, radius = options.radius;
    // coordinate
    chart.coordinate('theta', {
        innerRadius: innerRadius,
        radius: radius,
    });
    return params;
}
/**
 * statistic 配置
 * @param params
 */
function statistic(params) {
    var chart = params.chart, options = params.options;
    var statistic = options.statistic, percent = options.percent;
    // 先不处理 title 了，mini 应该没有 title 的需求了。
    var content = statistic.content;
    if (content) {
        var style = content.style, formatter = content.formatter, offsetX = content.offsetX, offsetY = content.offsetY, rotate = content.rotate;
        chart.annotation().text({
            position: ['50%', '50%'],
            content: formatter ? formatter({ percent: percent }) : percent,
            style: style,
            offsetX: offsetX,
            offsetY: offsetY,
            rotate: rotate,
        });
    }
    return params;
}
/**
 * 环形进度图适配器
 * @param chart
 * @param options
 */
function adaptor(params) {
    return utils_1.flow(adaptor_1.geometry, common_1.scale({}), coordinate, statistic, common_1.animation, common_1.theme, common_1.annotation())(params);
}
exports.adaptor = adaptor;
