// @antv/g2plot/esm/plots/funnel/geometries/basic.js
import { __rest } from "tslib";
import { deepMix, map } from '@antv/util';
import { isFunction } from '@antv/util';
import { flow, findGeometry } from '../../../utils';
import { FUNNEL_PERCENT } from '../constant';
import { transpose } from './util';
/**
 * 处理数据
 * @param params
 */
function format(params) {
    var options = params.options;
    var _a = options.data, data = _a === void 0 ? [] : _a, yField = options.yField;
    var formatData = [];
    // format 数据
    if (data[0][yField]) {
        formatData = map(data, function (row) {
            if (row[yField] !== undefined) {
                row[FUNNEL_PERCENT] = row[yField] / data[0][yField];
            }
            return row;
        });
    }
    return deepMix({}, params, {
        options: {
            formatData: formatData,
        },
    });
}
/**
 * geometry处理
 * @param params
 */
function geometry(params) {
    var chart = params.chart, options = params.options;
    var _a = options.formatData, formatData = _a === void 0 ? [] : _a, xField = options.xField, yField = options.yField, color = options.color;
    // 绘制漏斗图
    chart.data(formatData);
    chart
        .interval()
        .adjust('symmetric')
        .position(xField + "*" + yField + "*" + FUNNEL_PERCENT)
        .shape('funnel')
        .color(xField, color);
    return params;
}
/**
 * label 处理
 * @param params
 */
function label(params) {
    var chart = params.chart, options = params.options;
    var label = options.label, yField = options.yField, xField = options.xField;
    var geometry = findGeometry(chart, 'interval');
    if (!label) {
        geometry.label(false);
    }
    else {
        var callback = label.callback, cfg = __rest(label, ["callback"]);
        geometry.label({
            fields: [xField, yField, FUNNEL_PERCENT],
            callback: callback,
            cfg: cfg,
        });
    }
    return params;
}
/**
 * annotation 处理
 * @param params
 */
function annotation(params) {
    var chart = params.chart, options = params.options;
    var _a = options.formatData, formatData = _a === void 0 ? [] : _a, xField = options.xField, yField = options.yField, annotation = options.annotation;
    if (annotation !== false) {
        formatData.forEach(function (obj) {
            chart.annotation().text({
                top: true,
                position: [obj[xField], 'median'],
                content: isFunction(annotation) ? annotation(obj[xField], obj[yField], obj[FUNNEL_PERCENT], obj) : annotation,
                style: {
                    stroke: null,
                    fill: '#fff',
                    textAlign: 'center',
                },
            });
        });
    }
    return params;
}
/**
 * 基础漏斗
 * @param chart
 * @param options
 */
export function basicFunnel(params) {
    // flow 的方式处理所有的配置到 G2 API
    return flow(format, geometry, transpose, label, annotation)(params);
}
